File: AUGUST-2025/03-08-2025/p2106.scala
object Solution {
  def maxTotalFruits(fruits: Array[Array[Int]], startPos: Int, k: Int): Int = {
    def adjustLeft(l: Int, r: Int, sum: Int): (Int, Int) = {
      if (Math.min(startPos - 2 * fruits(l)(0) + fruits(r)(0),
        2 * fruits(r)(0) - fruits(l)(0) - startPos) > k) {
        adjustLeft(l + 1, r, sum - fruits(l)(1))
      } else {
        (l, sum)
      }
    }

    def collectFruits(l: Int, r: Int, sum: Int, maxSum: Int): Int = {
      if (r >= fruits.length || fruits(r)(0) > startPos + k) maxSum
      else {
        val updatedSum = sum + fruits(r)(1)
        val (adjustedL, adjustedSum) = adjustLeft(l, r, updatedSum)
        val updatedMaxSum = Math.max(maxSum, adjustedSum)
        collectFruits(adjustedL, r + 1, adjustedSum, updatedMaxSum)
      }
    }

    val initialL = fruits.indexWhere(fruit => fruit(0) >= startPos - k)
    if (initialL == -1) 0
    else collectFruits(initialL, initialL, 0, 0)
  }
}

File: AUGUST-2025/03-08-2025/p2106.java
class Solution {
    public int maxTotalFruits(int[][] fruits, int startPos, int k) {
        int left = 0;
        int right = 0;
        int n = fruits.length;
        int sum = 0;
        int max = 0;
        while (right < n) {
            sum += fruits[right][1];
            while (left <= right && !isValidRange(fruits[left][0], fruits[right][0], startPos, k)) {
                sum -= fruits[left][1];
                left++;
            }
            max = Math.max(max, sum);
            right++;
        }
        return max;
    }
    private boolean isValidRange(int leftPos, int rightPos, int startPos, int k) {
        if (rightPos <= startPos) {
            return startPos - leftPos <= k;
        } else if (leftPos >= startPos) {
            return rightPos - startPos <= k;
        } else {
            int option1 = (startPos - leftPos) * 2 + (rightPos - startPos);
            int option2 = (startPos - leftPos) + (rightPos - startPos) * 2;
            return Math.min(option1, option2) <= k;
        }
    }
}